### ISSUE TITLE
How to get AR for every class in mmdetection3.X?

### ISSUE THREAD

**Issue by THEBEST-cloud** [open-mmlab/mmdetection#10612]
sys.platform: win32
Python: 3.8.16 (default, Jun 12 2023, 21:00:42) [MSC v.1916 64 bit (AMD64)]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0: NVIDIA GeForce RTX 3060 Ti
CUDA_HOME: None
MSVC: 用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.36.32535 版
GCC: n/a
PyTorch: 1.8.0
PyTorch compiling details: PyTorch built with:
  - C++ Version: 199711
  - MSVC 192829337
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.7.0 (Git Hash 7aed236906b1f7a05c0917e5257a1af05e9ff683)
  - OpenMP 2019
  - CPU capability usage: AVX2
  - CUDA Runtime 11.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.0.5
  - Magma 2.5.4
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=C:/cb/pytorch_1000000000000/work/tmp_bin/sccache-cl.exe, CXX_FLAGS=/DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental -DNDEBUG -DUSE_FBGEMM -DUSE_XNNPACK, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.8.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=OFF, USE_OPENMP=ON, 

TorchVision: 0.9.0
OpenCV: 4.8.0
MMEngine: 0.8.0
mmdetection: 3.1.0+f78af77
---------------------------------------------------------------------------------------------------------------------------------------------------
How to get AR for every class in mmdetection3.X?

**Comment by GeorgePearse** [comment]
Tiny bit of help maybe, you can get AP for every class with 
```
cfg.val_evaluator.classwise = True
```

**Comment by da-orange-zi** [comment]
I have the same question , and I have set cfg.val_evaluator.classwise = True，but it didn't work ，please help！

**Comment by GeorgePearse** [comment]
@da-orange-zi  could you send more of your config?

**Comment by da-orange-zi** [comment]
Thank you for your reply@George Pearse, and this is my config
     auto_scale_lr = dict(base_batch_size=16, enable=False)
     backend_args = None
     data_root = '../data/coco/'
     dataset_type = 'CocoDataset'
     default_hooks = dict(
         checkpoint=dict(interval=20, type='CheckpointHook'),
         logger=dict(interval=20, type='LoggerHook'),
         param_scheduler=dict(type='ParamSchedulerHook'),
         sampler_seed=dict(type='DistSamplerSeedHook'),
         timer=dict(type='IterTimerHook'),
         visualization=dict(type='DetVisualizationHook'))
     default_scope = 'mmdet'
     env_cfg = dict(
         cudnn_benchmark=False,
         dist_cfg=dict(backend='nccl'),
         mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0))
     launcher = 'none'
     load_from = None
     log_level = 'INFO'
     log_processor = dict(by_epoch=True, type='LogProcessor', window_size=50)
     model = dict(
         backbone=dict(
             depth=50,
             frozen_stages=1,
             init_cfg=dict(checkpoint='torchvision://resnet50', type='Pretrained'),
             norm_cfg=dict(requires_grad=True, type='BN'),
             norm_eval=True,
             num_stages=4,
             out_indices=(
                 0,
                 1,
                 2,
                 3,
             ),
             style='pytorch',
             type='ResNet'),
         data_preprocessor=dict(
             bgr_to_rgb=True,
             mean=[
                 123.675,
                 116.28,
                 103.53,
             ],
            pad_size_divisor=32,
            std=[
                58.395,
                57.12,
                57.375,
            ],
            type='DetDataPreprocessor'),
        neck=dict(
            in_channels=[
                256,
                512,
                1024,
                2048,
            ],
            num_outs=5,
            out_channels=256,
            type='FPN'),
        roi_head=dict(
            bbox_head=dict(
                bbox_coder=dict(
                    target_means=[
                        0.0,
                        0.0,
                        0.0,
                        0.0,
                    ],
                    target_stds=[
                        0.1,
                        0.1,
                        0.2,
                        0.2,
                    ],
                    type='DeltaXYWHBBoxCoder'),
                fc_out_channels=1024,
                in_channels=256,
                loss_bbox=dict(loss_weight=1.0, type='L1Loss'),
                loss_cls=dict(
                    loss_weight=1.0, type='CrossEntropyLoss', use_sigmoid=False),
                num_classes=80,
                reg_class_agnostic=False,
                roi_feat_size=7,
                type='Shared2FCBBoxHead'),
            bbox_roi_extractor=dict(
                featmap_strides=[
                    4,
                    8,
                    16,
                    32,
                ],
                out_channels=256,
                roi_layer=dict(output_size=7, sampling_ratio=0, type='RoIAlign'),
                type='SingleRoIExtractor'),
            type='StandardRoIHead'),
        rpn_head=dict(
            anchor_generator=dict(
                ratios=[
                    0.5,
                    1.0,
                    2.0,
                ],
                scales=[
                    8,
                ],
                strides=[
                    4,
                    8,
                    16,
                    32,
                    64,
                ],
                type='AnchorGenerator'),
            bbox_coder=dict(
                target_means=[
                    0.0,
                    0.0,
                    0.0,
                    0.0,
                ],
                target_stds=[
                    1.0,
                    1.0,
                    1.0,
                    1.0,
                ],
                type='DeltaXYWHBBoxCoder'),
            feat_channels=256,
            in_channels=256,
            loss_bbox=dict(loss_weight=1.0, type='L1Loss'),
            loss_cls=dict(
                loss_weight=1.0, type='CrossEntropyLoss', use_sigmoid=True),
            type='RPNHead'),
        test_cfg=dict(
            rcnn=dict(
                max_per_img=100,
                nms=dict(iou_threshold=0.5, type='nms'),
                score_thr=0.05),
            rpn=dict(
                max_per_img=1000,
                min_bbox_size=0,
                nms=dict(iou_threshold=0.7, type='nms'),
                nms_pre=1000)),
        train_cfg=dict(
            rcnn=dict(
                assigner=dict(
                    ignore_iof_thr=-1,
                    match_low_quality=False,
                    min_pos_iou=0.5,
                    neg_iou_thr=0.5,
                    pos_iou_thr=0.5,
                    type='MaxIoUAssigner'),
                debug=False,
                pos_weight=-1,
                sampler=dict(
                    add_gt_as_proposals=True,
                    neg_pos_ub=-1,
                    num=512,
                    pos_fraction=0.25,
                    type='RandomSampler')),
            rpn=dict(
                allowed_border=-1,
                assigner=dict(
                    ignore_iof_thr=-1,
                    match_low_quality=True,
                    min_pos_iou=0.3,
                    neg_iou_thr=0.3,
                    pos_iou_thr=0.7,
                    type='MaxIoUAssigner'),
                debug=False,
                pos_weight=-1,
                sampler=dict(
                    add_gt_as_proposals=False,
                    neg_pos_ub=-1,
                    num=256,
                    pos_fraction=0.5,
                    type='RandomSampler')),
            rpn_proposal=dict(
                max_per_img=1000,
                min_bbox_size=0,
                nms=dict(iou_threshold=0.7, type='nms'),
                nms_pre=2000)),
        type='FasterRCNN')
    optim_wrapper = dict(
        optimizer=dict(lr=0.001, momentum=0.9, type='SGD', weight_decay=0.0001),
        type='OptimWrapper')
    param_scheduler = [
        dict(
            begin=0, by_epoch=False, end=500, start_factor=0.001, type='LinearLR'),
        dict(
            begin=0,
            by_epoch=True,
            end=12,
            gamma=0.1,
            milestones=[
                8,
                11,
            ],
            type='MultiStepLR'),
    ]
    resume = False
    test_cfg = dict(type='TestLoop')
    test_dataloader = dict(
        batch_size=1,
        dataset=dict(
            ann_file='annotations/instances_val2017.json',
            backend_args=None,
            data_prefix=dict(img='val2017/'),
            data_root='../data/coco/',
            pipeline=[
                dict(backend_args=None, type='LoadImageFromFile'),
                dict(keep_ratio=True, scale=(
                    1333,
                    800,
                ), type='Resize'),
                dict(type='LoadAnnotations', with_bbox=True),
                dict(
                    meta_keys=(
                        'img_id',
                        'img_path',
                        'ori_shape',
                        'img_shape',
                        'scale_factor',
                    ),
                    type='PackDetInputs'),
            ],
            test_mode=True,
            type='CocoDataset'),
        drop_last=False,
        num_workers=2,
        persistent_workers=True,
        sampler=dict(shuffle=False, type='DefaultSampler'))
    test_evaluator = dict(
        ann_file='../data/coco/annotations/instances_val2017.json',
        backend_args=None,
        format_only=False,
        metric='bbox',
        type='CocoMetric')
    test_pipeline = [
        dict(backend_args=None, type='LoadImageFromFile'),
        dict(keep_ratio=True, scale=(
            1333,
            800,
        ), type='Resize'),
        dict(type='LoadAnnotations', with_bbox=True),
        dict(
            meta_keys=(
                'img_id',
                'img_path',
                'ori_shape',
                'img_shape',
                'scale_factor',
            ),
            type='PackDetInputs'),
    ]
    train_cfg = dict(max_epochs=200, type='EpochBasedTrainLoop', val_interval=5)
    train_dataloader = dict(
        batch_sampler=dict(type='AspectRatioBatchSampler'),
        batch_size=2,
        dataset=dict(
            ann_file='annotations/instances_train2017.json',
            backend_args=None,
            data_prefix=dict(img='images/train/'),
            data_root='../data/coco/',
            filter_cfg=dict(filter_empty_gt=True, min_size=32),
            pipeline=[
                dict(backend_args=None, type='LoadImageFromFile'),
                dict(type='LoadAnnotations', with_bbox=True),
                dict(keep_ratio=True, scale=(
                    1333,
                    800,
                ), type='Resize'),
                dict(prob=0.5, type='RandomFlip'),
                dict(type='PackDetInputs'),
            ],
            type='CocoDataset'),
        num_workers=2,
        persistent_workers=True,
        sampler=dict(shuffle=True, type='DefaultSampler'))
    train_pipeline = [
        dict(backend_args=None, type='LoadImageFromFile'),
        dict(type='LoadAnnotations', with_bbox=True),
        dict(keep_ratio=True, scale=(
            1333,
            800,
        ), type='Resize'),
        dict(prob=0.5, type='RandomFlip'),
        dict(type='PackDetInputs'),
    ]
    val_cfg = dict(type='ValLoop')
    val_dataloader = dict(
        batch_size=1,
        dataset=dict(
            ann_file='annotations/instances_val2017.json',
            backend_args=None,
            data_prefix=dict(img='images/test/'),
            data_root='../data/coco/',
            pipeline=[
                dict(backend_args=None, type='LoadImageFromFile'),
                dict(keep_ratio=True, scale=(
                    1333,
                    800,
                ), type='Resize'),
                dict(type='LoadAnnotations', with_bbox=True),
                dict(
                    meta_keys=(
                        'img_id',
                        'img_path',
                        'ori_shape',
                        'img_shape',
                        'scale_factor',
                    ),
                    type='PackDetInputs'),
            ],
            test_mode=True,
            type='CocoDataset'),
        drop_last=False,
        num_workers=2,
        persistent_workers=True,
        sampler=dict(shuffle=False, type='DefaultSampler'))
    val_evaluator = dict(
        ann_file='../data/coco/annotations/instances_val2017.json',
        backend_args=None,
        format_only=False,
        metric='bbox',
        type='CocoMetric')
    vis_backends = [
        dict(type='LocalVisBackend'),
    ]
    visualizer = dict(
        name='visualizer',
        type='DetLocalVisualizer',
        vis_backends=[
            dict(type='LocalVisBackend'),
        ])
    work_dir = '/root/autodl-tmp/workdir'

**Comment by da-orange-zi** [comment]
![image](https://github.com/open-mmlab/mmdetection/assets/141907828/92d1cb30-97e0-45db-9b1a-8f7b6b5ba041)
![image](https://github.com/open-mmlab/mmdetection/assets/141907828/a204b462-3358-468d-ae36-d4e54b65c8cb)
I just get this,  just missing the per class results.
